src/Authoring/WinRT.SourceGenerator/Generator.cs

```diff
@@ -1,186 +1,226 @@
+            Logger.Log("Running " + cswinrtExe + " " + arguments);
+
+            var processInfo = new ProcessStartInfo
+            {
+                FileName = cswinrtExe,
+                Arguments = arguments,
+                UseShellExecute = false,
+                RedirectStandardOutput = true,
+                RedirectStandardError = true,
+                WindowStyle = ProcessWindowStyle.Hidden,
+                CreateNoWindow = true
+            };
+
+            try
+            {
+                using var cswinrtProcess = Process.Start(processInfo);
+                Logger.Log(cswinrtProcess.StandardOutput.ReadToEnd());
+                Logger.Log(cswinrtProcess.StandardError.ReadToEnd());
+                cswinrtProcess.WaitForExit();
+
+                if (cswinrtProcess.ExitCode != 0)
+                {
+                    throw new Win32Exception(cswinrtProcess.ExitCode);
+                }
+
+                foreach (var file in Directory.GetFiles(outputDir, "*.cs", SearchOption.TopDirectoryOnly))
+                {
+                    Logger.Log("Adding " + file);
+                    context.AddSource(Path.GetFileNameWithoutExtension(file), SourceText.From(File.ReadAllText(file), Encoding.UTF8));
+                }
+            }
+            finally
+            {
+                if (!GetKeepGeneratedSources(context))
+                {
+                    Directory.Delete(outputDir, true);
+                }
+            }
+        }
+
+        private string GetWinmdOutputFile(GeneratorExecutionContext context)
+        {
+            return Path.Combine(GetGeneratedFilesDir(context), GetAssemblyName(context) + ".winmd");
+        }
+
+        private void GenerateWinMD(MetadataBuilder metadataBuilder, string outputFile)
+        {
+            Logger.Log("Writing " + outputFile);
+            var managedPeBuilder = new ManagedPEBuilder(
+                new PEHeaderBuilder(
+                    machine: Machine.I386,
+                    imageCharacteristics: Characteristics.ExecutableImage | Characteristics.Dll | Characteristics.Bit32Machine),
+                new MetadataRootBuilder(metadataBuilder, "WindowsRuntime 1.4"),
+                new BlobBuilder(),
+                flags: CorFlags.ILOnly);
+
+            var peBlob = new BlobBuilder();
+            managedPeBuilder.Serialize(peBlob);
+
+            using var fs = new FileStream(outputFile, FileMode.Create, FileAccess.Write);
+            peBlob.WriteContentTo(fs);
+        }
+
+        public void Execute(GeneratorExecutionContext context)
+        {
+            if (!IsCsWinRTComponent(context))
+            {
+                return;
+            }
+
+            Logger.Initialize(context);
+
+            try
+            {
+                context.AddSource("System.Runtime.InteropServices.WindowsRuntime", SourceText.From(@"
+namespace System.Runtime.InteropServices.WindowsRuntime

```

Comment:
Do we want this as a private static string and just refer to it? it would be hard to adjust this in the future since it is kind of adhoc here. Also, just curious, did you try injecting this code in the Initialize function (instead of Execute) ?

Reply:
The latter is not supported today, but there is a proposed issue that I am tracking in the rosyln repo for this.  Ideally if that feature is added, I would move it to initialize as it makes the most sense there.